_posts/2016-09-04-build-maintainable-systems-with-hexagonal-architecture.md

```diff
@@ -0,0 +1,163 @@
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/service/LotteryServiceImpl.java?slice=33:"></script>
+
+`LotteryService` is the port and `LotteryServiceImpl` is the adapter. Notice how the implementation just delegates the request to `LotterySystem`.
+
+## Primary port for the administrators
+
+We also need to define the lottery administrator facing port. As before, we offer the relevant slice from `LotterySystem`.
+
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/administration/LotteryAdministration.java?slice=30:"></script>
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/administration/LotteryAdministrationImpl.java?slice=32:"></script>
+
+The implementation here is trivial since we can just delegate the requests to `LotterySystem`.
+
+## Secondary port for banking
+
+Next we implement the secondary ports and adapters. The first one is the banking support that enables us to manipulate bank account funds. To explain the concept, the player can write his bank account number on the lottery ticket and in case the ticket wins the lottery system automatically wire transfers the funds.
+
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/banking/WireTransfers.java?slice=23:"></script>
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/banking/WireTransfersImpl.java?slice=29:"></script>
+
+The bank back office is a simple `HashMap` based implementation. The lottery service's bank account is statically initialized to contain enough funds to pay the prizes in case some of the lottery tickets win.
+
+## Secondary port for notifications
+
+Another secondary port is the notification service. If the player has written his email address in the lottery ticket the system automatically sends notifications of the lottery system events, most importantly whether the ticket has won or not.
+
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/notifications/LotteryNotifications.java?slice=26:"></script>
+<script src="http://gist-it.appspot.com/http://github.com/iluwatar/java-design-patterns/raw/master/hexagonal/src/main/java/com/iluwatar/hexagonal/notifications/LotteryNotificationsImpl.java?slice=26:"></script>
+
+The methods in `LotteryNotificationsImpl` adapter are simple `System.out` printers so the implementation is trivial to understand.
```

Comment:
A better name for `LotteryNotificationsImpl` would be `ConsoleNotifications`. Also if the interface represents an entity that will send out notifications then the name can be something like `LotteryEventNotifier`, and the console implementation can have name `ConsoleNotifier` and maybe a `SMSNotifier`.
Also in the code there are `*Impl` classes, I am not fond of names like `*Impl`, rather the implementation should describe how it does, what it does. Plus why do classes like `LotterySystemImpl` hardcode their dependency on `LotteryNotificationsImpl`, `LotteryTicketInMemoryRepository`, `WireTransfersImpl`. Out of these lottery ticket repository and Wire Transfers are dependencies of the system and should be injected using constructor. What if we wanted to test the system using `MongoTicketRepository`?